docs/design/2020-06-01-global-kill.md

```diff
@@ -48,26 +54,67 @@ Bit 63 is always __ZERO__, making `connId` in range of non-negative int64, to be
+  - All `serverIds` within [2048, 2^22-1] when 64 bits `connId` chosen.
 
-On single TiDB instance without PD, a `serverId` of `1` is assigned.
+- On failure (e.g. fail connecting to PD, or all `serverId` are unavailable when 64 bits `connId` chosen), we block any new connection.
+- `serverId` is kept by PD with a lease (defaults to 12 hours, long enough to avoid brutally killing any long-run SQL). If TiDB is disconnected to PD longer than half of the lease (defaults to 6 hours), all connections are killed, and no new connection is accepted, to avoid running with stale/incorrect `serverId`. On connection to PD restored, a new `serverId` is acquired before accepting new connection.
+- On single TiDB instance without PD, a `serverId` of `1` is assigned.
 
 #### 5. local connId
-`local connId` is allocated by each TiDB instance on establishing connections incrementally.
+`local connId` is allocated by each TiDB instance on establishing connections auto-incrementally.
 
-Integer overflow is ignored at this stage, as `local connId` should be long enough.
+On system being busy and/or with long running SQL, `local connId` is possible to be integer-overflow and/or used up, especially on 32 bits `connId`. So we will check `local connId` existed or not before allocation (by maintaining a map of connection ids).
+a) In 32 bits `connId`, upgrade to 64 bits if `local connId` exhausted, or after retried more than 10 times (to avoid spending too much time in allocation).
+b) In 64 bits `connId`, return _"Too many connections"_ error if exhausted.
 
 #### 6. global kill
-On processing `KILL x` command, first extract `serverId` from `x`. Then if `serverId` aims to a remote TiDB instance, get the address from cluster info (see also [`CLUSTER_INFO`](https://docs.pingcap.com/tidb/stable/information-schema-cluster-info#cluster_info)), and redirect the command to it by "Coprocessor API" provided by the remote TiDB, along with the original user authentication.
+On processing `KILL x` command, first extract `serverId` from `x`. Then if `serverId` aims to a remote TiDB instance, get the address from [`CLUSTER_INFO`](https://docs.pingcap.com/tidb/stable/information-schema-cluster-info#cluster_info), and redirect the command to it by "Coprocessor API" provided by the remote TiDB, along with the original user authentication.
+
+#### 7. summary
+|      | 32 bits | 64 bits |
+| ---- | ---- | ---- |
+| ServerId pool size | 2^11 | 2^22 - 2^11 |
+| ServerId allocation | Random of __Unused__ serverIds acquired from PD within pool. Retry if unavailable. Upgrade to 64 bits while failed more than 3 times | Random of __All__ serverIds within pool. Retry if unavailable |
+| Local connId pool size | 2^20 | 2^40 |
+| Local connId allocation | Auto-increment within pool. Flip to zero when overflow. Upgrade to 64 bits while exhausted or retried more than 10 times | Auto-increment within pool. Flip to zero when overflow. Return "Too many connections" if exhausted |
```

Comment:
I would like to know what we will do in this case when new connection comes:

32bit

alive connection ids:
0, 1, 2, 3, 4, 5, ..., 1000, 2^20-1(last allocated id)

Reply:
```2^20 - 1 -> 0(failed #1) -> 1(failed #2) -> ... 9(failed #10) -> 2^20 (upgrade to 64 bits connId)```

I ever think about another solution about maintaining a recycle `local connId` queue (enqueue on client disconnect & dequeue on connect), which will consume maximum additional 8MB memory (2^20 x 4B), but have a better allocation performance & pool utilization.

En... the recycle queue solution seems to be better to me now. What do you think ? @breeswish

Reply:
lgtm

Reply:
Change design to queue solution. PTAL, thx~ @breeswish